File: Week 12/stacks.py
"""
Week 12 — Stacks
==================
Topics covered:
  - Stack class using a Python list
  - Balanced parentheses check
  - Next greater element (monotonic stack)
  - Min stack (O(1) push, pop, getMin)
  - Evaluate postfix (Reverse Polish Notation) expression

Each class/function includes time/space complexity analysis.
"""

from __future__ import annotations

from typing import List


# ===========================================================================
# Stack Implementation
# ===========================================================================

class Stack:
    """LIFO stack backed by a Python list.

    All operations are amortized O(1).
    """

    def __init__(self) -> None:
        self._data: list[int] = []

    def push(self, val: int) -> None:
        """Push *val* onto the stack.  O(1) amortized."""
        self._data.append(val)

    def pop(self) -> int:
        """Remove and return the top element.  Raises IndexError if empty."""
        if self.is_empty():
            raise IndexError("pop from empty stack")
        return self._data.pop()

    def peek(self) -> int:
        """Return the top element without removing it."""
        if self.is_empty():
            raise IndexError("peek from empty stack")
        return self._data[-1]

    def is_empty(self) -> bool:
        return len(self._data) == 0

    def size(self) -> int:
        return len(self._data)

    def __repr__(self) -> str:
        return f"Stack({self._data})"


# ===========================================================================
# Balanced Parentheses
# ===========================================================================

def is_balanced(s: str) -> bool:
    """Check if a string of brackets is balanced.

    Supports '()', '[]', '{}'.

    Time:  O(n)
    Space: O(n)  — the stack
    """
    matching = {")": "(", "]": "[", "}": "{"}
    stack: list[str] = []

    for ch in s:
        if ch in "([{":
            stack.append(ch)
        elif ch in ")]}":
            if not stack or stack[-1] != matching[ch]:
                return False
            stack.pop()

    return len(stack) == 0


# ===========================================================================
# Next Greater Element (Monotonic Stack)
# ===========================================================================

def next_greater_element(arr: List[int]) -> List[int]:
    """For each element, find the next element to the right that is greater.

    Returns a list where result[i] is the next greater element for arr[i],
    or -1 if none exists.

    Uses a monotonic decreasing stack: we process elements left to right
    and pop elements from the stack that are smaller than the current
    element (the current element is their "next greater").

    Time:  O(n) — each element is pushed and popped at most once
    Space: O(n)
    """
    n = len(arr)
    result = [-1] * n
    stack: list[int] = []  # stores indices

    for i in range(n):
        while stack and arr[stack[-1]] < arr[i]:
            idx = stack.pop()
            result[idx] = arr[i]
        stack.append(i)

    return result


# ===========================================================================
# Min Stack
# ===========================================================================

class MinStack:
    """A stack that supports push, pop, top, and retrieving the minimum
    element — all in O(1) time.

    We maintain a parallel stack that tracks the current minimum at each
    level of the main stack.

    Space: O(n)
    """

    def __init__(self) -> None:
        self._stack: list[int] = []
        self._min_stack: list[int] = []

    def push(self, val: int) -> None:
        """Push *val* and update the min tracker.  O(1)"""
        self._stack.append(val)
        current_min = min(val, self._min_stack[-1]) if self._min_stack else val
        self._min_stack.append(current_min)

    def pop(self) -> int:
        """Pop and return the top element.  O(1)"""
        if not self._stack:
            raise IndexError("pop from empty MinStack")
        self._min_stack.pop()
        return self._stack.pop()

    def top(self) -> int:
        """Return the top element without removing it.  O(1)"""
        if not self._stack:
            raise IndexError("top from empty MinStack")
        return self._stack[-1]

    def get_min(self) -> int:
        """Return the minimum element in the stack.  O(1)"""
        if not self._min_stack:
            raise IndexError("get_min from empty MinStack")
        return self._min_stack[-1]


# ===========================================================================
# Evaluate Postfix Expression
# ===========================================================================

def eval_postfix(tokens: List[str]) -> int:
    """Evaluate a postfix (Reverse Polish Notation) expression.

    Supported operators: +, -, *, /  (integer division truncated toward zero).

    Time:  O(n)
    Space: O(n)

    Example:
        ["2", "1", "+", "3", "*"]  →  9   because (2+1)*3 = 9
    """
    stack: list[int] = []

    for token in tokens:
        if token in ("+", "-", "*", "/"):
            b = stack.pop()
            a = stack.pop()
            if token == "+":
                stack.append(a + b)
            elif token == "-":
                stack.append(a - b)
            elif token == "*":
                stack.append(a * b)
            else:
                # Truncate toward zero (Python's // rounds toward -inf)
                stack.append(int(a / b))
        else:
            stack.append(int(token))

    return stack[0]


# ---------------------------------------------------------------------------
# Tests
# ---------------------------------------------------------------------------

if __name__ == "__main__":
    # ---- Stack class ----
    s = Stack()
    s.push(10)
    s.push(20)
    s.push(30)
    assert s.peek() == 30
    assert s.pop() == 30
    assert s.size() == 2
    assert s.is_empty() is False
    print("[PASS] Stack class")

    # ---- Balanced Parentheses ----
    assert is_balanced("()[]{}") is True
    assert is_balanced("([{}])") is True
    assert is_balanced("(]") is False
    assert is_balanced("([)]") is False
    assert is_balanced("") is True
    assert is_balanced("{[()]}") is True
    print("[PASS] Balanced parentheses")

    # ---- Next Greater Element ----
    assert next_greater_element([4, 5, 2, 25]) == [5, 25, 25, -1]
    assert next_greater_element([13, 7, 6, 12]) == [-1, 12, 12, -1]
    assert next_greater_element([1, 2, 3, 4]) == [2, 3, 4, -1]
    assert next_greater_element([4, 3, 2, 1]) == [-1, -1, -1, -1]
    print("[PASS] Next greater element (monotonic stack)")

    # ---- Min Stack ----
    ms = MinStack()
    ms.push(-2)
    ms.push(0)
    ms.push(-3)
    assert ms.get_min() == -3
    ms.pop()
    assert ms.top() == 0
    assert ms.get_min() == -2
    ms.push(-1)
    assert ms.get_min() == -2
    print("[PASS] Min stack")

    # ---- Evaluate Postfix ----
    assert eval_postfix(["2", "1", "+", "3", "*"]) == 9
    assert eval_postfix(["4", "13", "5", "/", "+"]) == 6
    assert eval_postfix(["10", "6", "9", "3", "+",
                          "-11", "*", "/", "*", "17", "+", "5", "+"]) == 22
    print("[PASS] Evaluate postfix expression")

    print("\nAll Week 12 tests passed!")

File: Week 12/rust/stacks.rs
//! # Week 12: Stacks
//!
//! This module covers stack data structures and classic stack-based algorithms in Rust.
//! Topics include:
//! - Stack implementation using `Vec<i32>`
//! - Balanced parentheses check
//! - Next greater element (monotonic stack)
//! - MinStack (stack with O(1) minimum query)
//! - Postfix expression evaluation
//!
//! ## Rust-Specific Notes for DSA Learners
//! - `Vec<T>` is the natural stack in Rust: `push()` and `pop()` are O(1) amortized.
//! - `pop()` returns `Option<T>` — Rust forces you to handle the empty-stack case.
//!   No null pointer exceptions or silent undefined behavior.
//! - `last()` / `peek()` returns `Option<&T>` — a borrowed reference, not a copy.
//!   This means the stack still owns the element.

use std::collections::HashMap;

// ===========================================================================
// Stack Implementation
// ===========================================================================

/// A simple stack backed by a `Vec<i32>`.
///
/// In practice, you'd use `Vec<T>` directly — this wrapper exists to demonstrate
/// the stack interface explicitly.
struct Stack {
    data: Vec<i32>,
}

impl Stack {
    fn new() -> Self {
        Stack { data: Vec::new() }
    }

    /// Pushes a value onto the top of the stack.
    /// # Complexity: O(1) amortized
    fn push(&mut self, value: i32) {
        self.data.push(value);
    }

    /// Removes and returns the top element, or `None` if empty.
    /// # Complexity: O(1)
    fn pop(&mut self) -> Option<i32> {
        self.data.pop()
    }

    /// Returns a reference to the top element without removing it.
    /// # Complexity: O(1)
    fn peek(&self) -> Option<&i32> {
        self.data.last()
    }

    /// Returns true if the stack is empty.
    fn is_empty(&self) -> bool {
        self.data.is_empty()
    }

    /// Returns the number of elements.
    fn len(&self) -> usize {
        self.data.len()
    }
}

// ===========================================================================
// Balanced Parentheses
// ===========================================================================

/// Checks if a string of brackets is balanced.
///
/// Supports `()`, `[]`, `{}`. Uses a stack: push opening brackets, pop and
/// compare when we see a closing bracket.
///
/// # Complexity
/// - Time:  O(n)
/// - Space: O(n) worst case (all opening brackets)
fn is_balanced(s: &str) -> bool {
    let mut stack: Vec<char> = Vec::new();

    // Map closing brackets to their matching opening bracket.
    let matches: HashMap<char, char> = [(')', '('), (']', '['), ('}', '{')]
        .iter()
        .cloned()
        .collect();

    for ch in s.chars() {
        match ch {
            '(' | '[' | '{' => stack.push(ch),
            ')' | ']' | '}' => {
                // `matches.get(&ch)` returns Option<&char>.
                // We check that the top of the stack matches.
                match (stack.pop(), matches.get(&ch)) {
                    (Some(top), Some(&expected)) if top == expected => {}
                    _ => return false,
                }
            }
            _ => {} // Ignore non-bracket characters.
        }
    }

    stack.is_empty() // Must be empty — all brackets matched.
}

// ===========================================================================
// Next Greater Element (Monotonic Stack)
// ===========================================================================

/// For each element in the array, finds the next greater element to its right.
///
/// Uses a **monotonic decreasing stack**: we maintain a stack of indices whose
/// values haven't found their "next greater" yet. When we encounter a larger
/// element, we pop and assign.
///
/// Returns a `Vec<Option<i32>>`: `Some(value)` if a greater element exists,
/// `None` if not (Rust's `Option` is perfect for this).
///
/// # Complexity
/// - Time:  O(n) — each element is pushed and popped at most once
/// - Space: O(n)
fn next_greater_element(arr: &[i32]) -> Vec<Option<i32>> {
    let n = arr.len();
    let mut result: Vec<Option<i32>> = vec![None; n];
    let mut stack: Vec<usize> = Vec::new(); // Stack of indices

    for i in 0..n {
        // Pop all elements that are smaller than arr[i].
        while let Some(&top_idx) = stack.last() {
            if arr[top_idx] < arr[i] {
                stack.pop();
                result[top_idx] = Some(arr[i]);
            } else {
                break;
            }
        }
        stack.push(i);
    }
    // Elements remaining in the stack have no next greater element (already None).

    result
}

// ===========================================================================
// MinStack
// ===========================================================================

/// A stack that supports push, pop, top, and retrieving the minimum element
/// in O(1) time.
///
/// ## Strategy
/// We maintain two stacks in parallel:
/// - `data`: the actual elements.
/// - `mins`: tracks the minimum at each level. `mins[i]` is the minimum of
///   all elements from the bottom up to position `i`.
///
/// This uses O(n) extra space but guarantees O(1) `get_min()`.
///
/// # Complexity (all operations)
/// - Time:  O(1)
/// - Space: O(n) total
struct MinStack {
    data: Vec<i32>,
    mins: Vec<i32>,
}

impl MinStack {
    fn new() -> Self {
        MinStack {
            data: Vec::new(),
            mins: Vec::new(),
        }
    }

    /// Pushes a value and updates the running minimum.
    fn push(&mut self, value: i32) {
        self.data.push(value);
        // The new minimum is min(value, current_min).
        let new_min = match self.mins.last() {
            Some(&current_min) => value.min(current_min),
            None => value,
        };
        self.mins.push(new_min);
    }

    /// Pops the top element (also pops from the mins stack).
    fn pop(&mut self) -> Option<i32> {
        self.mins.pop();
        self.data.pop()
    }

    /// Returns the top element.
    fn top(&self) -> Option<&i32> {
        self.data.last()
    }

    /// Returns the current minimum in O(1).
    fn get_min(&self) -> Option<&i32> {
        self.mins.last()
    }

    fn is_empty(&self) -> bool {
        self.data.is_empty()
    }
}

// ===========================================================================
// Evaluate Postfix Expression
// ===========================================================================

/// Evaluates a postfix (Reverse Polish Notation) expression.
///
/// Tokens are space-separated. Operands are pushed onto a stack; operators pop
/// two operands, compute the result, and push it back.
///
/// Supported operators: `+`, `-`, `*`, `/` (integer division, truncates toward zero).
///
/// Returns `Some(result)` on success, `None` on invalid expression.
///
/// # Complexity
/// - Time:  O(n) where n = number of tokens
/// - Space: O(n) for the stack
fn eval_postfix(expression: &str) -> Option<i32> {
    let mut stack: Vec<i32> = Vec::new();

    for token in expression.split_whitespace() {
        match token {
            "+" | "-" | "*" | "/" => {
                // Need at least two operands.
                let b = stack.pop()?;
                let a = stack.pop()?;
                let result = match token {
                    "+" => a + b,
                    "-" => a - b,
                    "*" => a * b,
                    "/" => {
                        if b == 0 {
                            return None; // Division by zero.
                        }
                        a / b // Integer division (truncates toward zero in Rust).
                    }
                    _ => unreachable!(),
                };
                stack.push(result);
            }
            _ => {
                // Try to parse as integer.
                let num: i32 = token.parse().ok()?;
                stack.push(num);
            }
        }
    }

    // Valid postfix expression leaves exactly one value on the stack.
    if stack.len() == 1 {
        stack.pop()
    } else {
        None
    }
}

// ===========================================================================
// Main — demonstrations and test assertions
// ===========================================================================

fn main() {
    println!("=== Week 12: Stacks ===\n");

    // --- Stack Implementation ---
    println!("--- Stack (Vec-backed) ---");
    let mut stack = Stack::new();
    assert!(stack.is_empty());

    stack.push(10);
    stack.push(20);
    stack.push(30);
    assert_eq!(stack.len(), 3);
    assert_eq!(stack.peek(), Some(&30));
    println!("Stack after push(10, 20, 30): top = {:?}", stack.peek());

    assert_eq!(stack.pop(), Some(30));
    assert_eq!(stack.pop(), Some(20));
    assert_eq!(stack.peek(), Some(&10));
    println!("After two pops: top = {:?}", stack.peek());

    assert_eq!(stack.pop(), Some(10));
    assert_eq!(stack.pop(), None); // Empty!
    println!("Pop from empty stack: {:?}", stack.pop());

    // --- Balanced Parentheses ---
    println!("\n--- Balanced Parentheses ---");
    assert!(is_balanced("()"));
    assert!(is_balanced("()[]{}"));
    assert!(is_balanced("{[()]}"));
    assert!(is_balanced(""));
    assert!(!is_balanced("(]"));
    assert!(!is_balanced("([)]"));
    assert!(!is_balanced("(("));
    assert!(!is_balanced(")"));
    assert!(is_balanced("(a + b) * [c - {d / e}]")); // Non-bracket chars ignored
    println!("is_balanced(\"{{[()]}}\") = {}", is_balanced("{[()]}"));
    println!("is_balanced(\"([)]\")    = {}", is_balanced("([)]"));

    // --- Next Greater Element ---
    println!("\n--- Next Greater Element (Monotonic Stack) ---");
    let arr = vec![4, 5, 2, 10, 8];
    let nge = next_greater_element(&arr);
    assert_eq!(nge, vec![Some(5), Some(10), Some(10), None, None]);
    println!("arr = {:?}", arr);
    println!("NGE = {:?}", nge);

    let arr2 = vec![3, 2, 1];
    let nge2 = next_greater_element(&arr2);
    assert_eq!(nge2, vec![None, None, None]);
    println!("arr = {:?}, NGE = {:?}", arr2, nge2);

    let arr3 = vec![1, 2, 3];
    let nge3 = next_greater_element(&arr3);
    assert_eq!(nge3, vec![Some(2), Some(3), None]);
    println!("arr = {:?}, NGE = {:?}", arr3, nge3);

    // --- MinStack ---
    println!("\n--- MinStack ---");
    let mut ms = MinStack::new();
    ms.push(5);
    ms.push(3);
    ms.push(7);
    ms.push(2);
    ms.push(8);
    assert_eq!(ms.get_min(), Some(&2));
    assert_eq!(ms.top(), Some(&8));
    println!("Stack: [5, 3, 7, 2, 8], min = {:?}, top = {:?}", ms.get_min(), ms.top());

    ms.pop(); // Remove 8
    assert_eq!(ms.get_min(), Some(&2));
    println!("After pop(8): min = {:?}", ms.get_min());

    ms.pop(); // Remove 2
    assert_eq!(ms.get_min(), Some(&3));
    println!("After pop(2): min = {:?}", ms.get_min());

    ms.pop(); // Remove 7
    assert_eq!(ms.get_min(), Some(&3));
    println!("After pop(7): min = {:?}", ms.get_min());

    ms.pop(); // Remove 3
    assert_eq!(ms.get_min(), Some(&5));
    println!("After pop(3): min = {:?}", ms.get_min());

    // --- Evaluate Postfix ---
    println!("\n--- Evaluate Postfix Expression ---");
    // "3 4 + 2 * 7 /" = ((3 + 4) * 2) / 7 = 14 / 7 = 2
    assert_eq!(eval_postfix("3 4 + 2 * 7 /"), Some(2));
    println!("\"3 4 + 2 * 7 /\" = {:?}", eval_postfix("3 4 + 2 * 7 /"));

    // "5 1 2 + 4 * + 3 -" = 5 + ((1 + 2) * 4) - 3 = 5 + 12 - 3 = 14
    assert_eq!(eval_postfix("5 1 2 + 4 * + 3 -"), Some(14));
    println!("\"5 1 2 + 4 * + 3 -\" = {:?}", eval_postfix("5 1 2 + 4 * + 3 -"));

    // Simple: "2 3 +"
    assert_eq!(eval_postfix("2 3 +"), Some(5));

    // Division truncates: "7 2 /"
    assert_eq!(eval_postfix("7 2 /"), Some(3));

    // Invalid expression
    assert_eq!(eval_postfix("1 2 3 +"), None); // Two values left on stack
    assert_eq!(eval_postfix("+"), None);         // Not enough operands

    // Negative numbers
    assert_eq!(eval_postfix("-3 4 +"), Some(1));
    println!("\"-3 4 +\" = {:?}", eval_postfix("-3 4 +"));

    println!("\nAll assertions passed!");
}
